src/cmdline/mod.rs

```diff
@@ -345,42 +290,216 @@ impl Cmdline {
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new().unwrap();
+    /// assert!(cl.boot_args_mut().insert_key_value("a", "b").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "a=b");
+    /// ```
+    pub fn boot_args_mut(&mut self) -> &mut BootArgs {
+        &mut self.boot_args
+    }
+
+    /// Returns the InitArgs representation of the command line.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new().unwrap();
+    /// assert!(cl.boot_args_mut().insert_key_value("a", "b").is_ok());
+    /// assert!(cl.init_args_mut().insert_key_value("c", "d").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "a=b -- c=d");
+    /// ```
+    pub fn init_args_mut(&mut self) -> &mut InitArgs {
+        &mut self.init_args
+    }
+
+    /// Returns a C compatible representation of the command line
+    /// The Linux kernel expects a null terminated cmdline according to the source:
+    /// https://elixir.bootlin.com/linux/v5.10.139/source/kernel/params.c#L179
+    ///
+    /// To get bytes of the cmdline to be written in guest's memory (including the
+    /// null terminator) from this representation, use CString::as_bytes_with_nul()
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new();
+    /// cl.insert_string("foobar");
+    /// assert_eq!(cl.as_cstring().unwrap().as_bytes_with_nul(), b"foobar\0");
+    /// ```
+    pub fn as_cstring(&self) -> Result<CString> {
+        if self.boot_args.0.is_empty() {
+            return if self.init_args.0.is_empty() {
+                CString::new("".to_string()).map_err(|_| Error::NullTerminator)
+            } else {
+                Err(Error::NoBootArgsInserted)
+            }
+        }
+
+        if self.init_args.0.is_empty() {
+            if self.boot_args.0.len() + 1 > CMDLINE_MAXSIZE {
+                // need space for the null terminator
+                Err(Error::TooLarge)
+            } else {
+                CString::new(self.boot_args.0.to_string()).map_err(|_| Error::NullTerminator)
+            }
+        } else if self.boot_args.0.len() + self.init_args.0.len() + 5 > CMDLINE_MAXSIZE {
+            // need space for the null terminator and ' -- ' delimiter
+            Err(Error::TooLarge)
+        } else {
+            CString::new(format!("{} -- {}", self.boot_args.0, self.init_args.0)).map_err(|_| Error::NullTerminator)
+        }
+    }
+}
+
+impl TryFrom<String> for Cmdline {
+    type Error = String;
+
+    fn try_from(cmdline_raw: String) -> result::Result<Self, Self::Error> {
+        // Checking for ' -- ' sequence is required as long as a sequence like 'foo-- bar' should not be
+        // handled as a boot param 'foo' and an init arg 'bar'
+        let (mut boot_args, mut init_args) = match cmdline_raw.split_once(" -- ") {
```

Comment:
Can we have a define for " -- " instead? For example, we can call it INIT_ARG_DELIMITER. Then, we should be able to use INIT_ARG_DELIMITER.len() instead of magic numbers in the code below.

Reply:
We already have such a define: INIT_ARGS_SEPARATOR; I don't see any magic number below apart some variations from the ' -- ' separator such as '-- ' or ' --' that are used to catch some corner cases (but their only usecase is in try_from() so I considered that is easier to see the actual value instead of another define which is used in only one place). Looks like this comment is older and I forgot to close it or something... ping me if I miss something